---
title: Authors
feature_text: |
  ## RNA-seq Bioinformatics
  Introduction to bioinformatics for RNA sequence analysis
feature_image: "assets/genvis-dna-bg_optimized_v1a.png"
excerpt: ""
layout: "home"
---

***

### Malachi Griffith
<nav class="nav  nav--social">
{% for link in site.contact_malachi %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/MalachiGriffith.png" position="right" class="rounded-circle" width="300" %}
Dr. Malachi Griffith is an Assistant Director of the McDonnell Genome Institute and Professor in the Department of Medicine. Dr. Griffith's research is focused on the development of genomics, bioinformatics and statistical methods as they apply to the study of cancer biology and medicine. A particular focus of his work is in the translation of genomics data from whole genome, exome and transcriptome sequencing into clinically actionable observations and personalized cancer therapies.
 Dr. Griffith is a regular instructor for the Canadian Bioinformatics Workshops series, Cold Spring Harbor Laboratories and Cancer Research Institute. Before coming to Washington University, Dr. Griffith was a bioinformatics post-doctoral fellow at the BC Cancer Agency Genome Sciences center in Vancouver, British Columbia. He received his Ph.D. (Medical Genetics, 2009) from the University of British Columbia and B.S. (Biochemistry and Biology with Honors, 2002) from the University of Winnipeg.

***

### Obi L. Griffith
<nav class="nav  nav--social">
{% for link in site.contact_obi %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/ObiGriffith.png" position="right" class="rounded-circle" width="300" %}
Dr. Obi Griffith is Professor of Medicine and Assistant Director at the McDonnell Genome Institute. Dr. Griffith's research is focused on the development of personalized medicine strategies for cancer using genomic technologies with a particular focus on gene regulatory changes associated with breast cancer. He develops and uses bioinformatics and statistical methods for the analysis of high throughput sequence data and identification of biomarkers for diagnostic, prognostic and drug response prediction. Dr. Griffith has developed and instructs a workshop on RNA sequence analysis for Cold Spring Harbor Laboratories and is a regular instructor for the Canadian Bioinformatics Workshops series and Cancer Research Insitute. Before coming to Washington University, Dr. Griffith completed bioinformatics post-doctoral fellowships at Lawrence Berkeley National Laboratory in Berkeley, California and at the BC Cancer Agency Genome Sciences Centre in Vancouver, British Columbia. He received his Ph.D. (Medical Genetics, 2008) from the University of British Columbia and B.S. (Biochemistry and Biology with Honors, 2002) from the University of Winnipeg.

***

### Brian Haas
<nav class="nav  nav--social">
{% for link in site.contact_brian %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/BrianHaas.png" position="right" class="rounded-circle" width="300" %}

Brian Haas earned a master’s degree in molecular biology and biochemistry at the State University of Albany, N.Y., and after discovering his passion for deciphering the genetic blueprint of life, decided to enter the then-burgeoning field of bioinformatics. He has since been pursuing his dream of doing bioinformatics and genome research, along the way earning a separate master’s degree in computer science at Johns Hopkins University. Brian has been working in the genomics field since 1999, first with The Institute for Genomic Research in Rockville, Md. He joined the Broad Institute in 2007 and is currently a senior computational biologist within the Klarman Cell Observatory. His research is focused on developing and supporting open source bioinformatics software tools and using computational techniques to explore aspects of molecular biology, particularly in the study of transcriptomes.

***


### Allegra Petti
<nav class="nav  nav--social">
{% for link in site.contact_allegra %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/AllegraPetti.png" position="right" class="rounded-circle" width="300" %}

Allegra Petti is an Assistant Professor in the Department of Medicine, Division of Oncology, and the McDonnell Genome Institute at Washington University School of Medicine. Her research focuses on gene regulation during the development and evolution of cancer. She       develops analytical approaches for the integration and interpretation of multidimensional genomic data, with an emphasis on single-cell RNA-sequencing (scRNA-seq) data. Dr. Petti’s current research is aimed at deciphering the relationships among genetic,                 transcriptional, and microenvironmental heterogeneity in AML and solid tumors. Before coming to Washington University, Dr. Petti was a post-doctoral fellow with Prof. David Botstein at Princeton University. She earned a PhD from Harvard Medical School in the lab of      Prof. George Church, and received a BA in Biology from Rice University.

***

### Aaron Quinlan
<nav class="nav  nav--social">
{% for link in site.contact_aaron %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/AaronQuinlan.png" position="right" class="rounded-circle" width="300" %}

Aaron Quinlan is a Professor in the Departments of Human Genetics and Biomedical Informatics at the University of Utah. He received his bachelor’s degree in Computer Science from the College of William and Mary and his Ph.D. from Boston College where he focused on population genetics, new methods for emerging DNA sequencing technologies, and the discovery and characterization of genetic variation. During his NRSA postdoctoral fellowship at the University of Virginia he developed expertise in structural variation of mammalian genomes and somatic genome mutation. He started his laboratory at the University of Virginia in 2011. He was recruited to the University of Utah in early 2015 to become the Associate Director of the Utah Center for Genetic Discovery. Broadly speaking, the Quinlan laboratory is interested in the development and application of new computational and statistical techniques for understanding the biology of genomes. His team tackles problems with practical importance to identifying genome variation, understanding genome evolution, and mining genetic variation underlying rare genetic disease. The Quinlan laboratory’s ultimate goal is to develop and apply computational technologies that improve our understanding of human disease.

***

### Jason Walker
<nav class="nav  nav--social">
{% for link in site.contact_jason %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/JasonWalker.png" position="right" class="rounded-circle" width="300" %}
Jason Walker is a Biomedical Informatics Project Manager at the McDonnell Genome Institute. He works with the Griffith Lab to translate grant objectives and analysis goals into well defined development milestones for software engineers to iterate on projects, achieving consistent progress while maintaining best practices in software development methodologies. His responsibilities include overseeing the integration and adaptation of numerous bioinformatics tools for cancer genomics automated analysis of high-throughput DNA and RNA sequence data. In addition to large-scale research projects, he works with a CLIA licensed laboratory on clinical-grade, CAP-accredited analysis workflows for reporting somatic variants in cancer patient genomes.'

***

### Alex Wagner
<nav class="nav  nav--social">
{% for link in site.contact_alex %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/AlexWagner.png" position="right" class="rounded-circle" width="300" %}
Dr. Wagner was an <a href="https://www.cancer.gov/grants-training/training/funding/f32">NCI Ruth L. Kirschstein NRSA Postdoctoral Fellow</a> at the McDonnell Genome Institute and Washington University School of Medicine. Dr. Wagner now leads his own lab at Nationwide Children's Hospital. His research interests are in facilitating the interpretation of genomic variants in a clinical context. He leads development of the <a href="http://dgidb.org">Drug Gene Interaction Database</a>, and is a primary contributor to other projects centered on interpretations of genomic variants (<a href="http://docm.info">Database of Curated Mutations</a>, <a href="http://civicdb.org">Clinical Interpretations of Variants in Cancer</a>) and <a href="https://bioconductor.org/packages/release/bioc/html/GenVisR.html">Genomic Visualizations in R</a>. Alex is an active member of the <a href="http://genomicsandhealth.org/">Global Alliance for Genomics and Health</a> (GA4GH), and is leading an effort to aggregate and normalize the disparate knowledgebases of the GA4GH <a href="http://cancervariants.org">Variant Interpretation for Cancer Consortium</a>.

***

### Kelsy C. Cotto
<nav class="nav  nav--social">
{% for link in site.contact_kelsy %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/KelsyCotto.png" position="right" class="rounded-circle" width="300" %}
Kelsy was a PhD candidate in the Molecular Cell Biology program at Washington University in St. Louis. She completed her undergraduate degree at Mercer University in 2016, where she earned a B.S. in Biochemistry and Molecular Biology. She is interested in developing methods to analyze multiple types of sequencing data in order to better understand regulatory mutations and splicing within cancer, particularly with respect to personalized cancer vaccine design. Dr. Cotto now work in the biotech industry in San Francisco.

***

### Zachary L. Skidmore
<nav class="nav  nav--social">
{% for link in site.contact_zach %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/ZacharySkidmore.png" position="right" class="rounded-circle" width="300" %}
Zachary Skidmore is a staff scientist at the McDonell Genome Institute at Washington University in Saint Louis. His undergraduate work was completed at the Ohio State University where he obtained a B.Sc. in molecular biology. Graduate work was performed at the University of Illinois where he obtained a M.eng in bioinformatics. His research focus is in the realm of cancer biology where he uses and develops tools and techniques to aid in the analysis and interpretation of cancer sequencing data. He has worked on several large sequencing projects across many cancer types and has expertise in a variety of languages (perl, R, bash, python, typescript, angular2). He is the creator and maintainer of the bioconductor package GenVisR, a graphics program designed to help visualize cohort level genomic data.

***

### Chris A. Miller
<nav class="nav  nav--social">
{% for link in site.contact_chris %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/ChrisMiller.png" position="right" class="rounded-circle" width="300" %}
Chris Miller is an Associate Professor in the Division of Oncology and a member of the McDonnell Genome Institute. His focus is on developing and applying computational tools to provide insight into the origins and progression of cancer. Dr. Miller received Bachelor degrees in Biology and Computer Science from Truman State University. He received his PhD in Computational Biology from Baylor College of Medicine. His core research interests include understanding the clonal architecture of tumors and how they evolve in response to therapy, with a special focus on hematologic cancers. His research also focuses on effective designing and targeting of cancer immunotherapies, developing open-source software for interpreting and visualizing genomic data, and integrative analysis that translates multi-dimensional genomic data into both functional and actionable contexts.

*** 

### Huiming Xia
<nav class="nav  nav--social">
{% for link in site.contact_huiming %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/HuimingXia.png" position="right" class="rounded-circle" width="300"%}
Huiming Xia was a PhD Candidate in the Computational Systems & Biology program at Washington University in St. Louis. She graduated from Carnegie Mellon University in 2017 with a dual degree in Computational Biology and Chemistry. Dr. Xia is interested in developing machine learning models for improving current strategies of neoantigen design and prioritization for cancer immunotherapy. Dr. Xia now works for Moderna.<br><br><br><br>

***

### Saad Khan
<nav class="nav  nav--social">
{% for link in site.contact_saad %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/SaadKhan.png" position="right" class="rounded-circle" width="300"%}
Saad Khan is a Bioinformatics Scientist at Washington University in St. Louis. He graduated with Ph.D in Informatics from University of Missouri-Columbia in December 2019. His current work focuses on understanding Acute Myeloid Leukemia at the single cell level using Single cell RNA-sequencing analysis. He is interested in developing informatics approaches in Single cell RNA-seq to better understand cancer. <br><br><br><br>

***

### Miller Richters
<nav class="nav  nav--social">
{% for link in site.contact_megan %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/MeganRichters.png" position="right" class="rounded-circle" width="300"%}
Miller Richters was a PhD Candidate in the Molecular Genetics and Genomics program at Washington University in St. Louis and now works for Natera, Inc. Miller graduated from the University of Louisiana at Monroe in 2014 with a B.S. in Biology. Dr. Miller is interested in assessing the impact of intratumoral heterogeneity on neoantigen prediction accuracy as well as exploring novel sources of neoantigens for personalized cancer vaccines. <br><br><br><br>

***

### Felicia Gomez
<nav class="nav  nav--social">
{% for link in site.contact_felicia %}
    {% assign id = link[0] | downcase %}
    {% capture refer %}{{link[1]}}{% endcapture %}
<a class="link" target="_blank" href="{{refer}}" title="{{link[0]}}">{% include icon.html id=id %}</a>
{% endfor %}
</nav>
{% include figure.html image="/assets/Authors/FeliciaGomez.png" position="right" class="rounded-circle" width="300"%}
Felicia Gomez is an Assistant Professor of Medicine at Washington University in St. Louis. She is leading a research program focused on lymphoma genomics with the goal of describing somatic events that are characteristic of this malignancy. She also collaborates with members of the Griffith and Fehniger laboratories, as well was the Washington University Genomics Tumor Board. <br><br><br><br>

***


