---
title: RNAseq
feature_text: |
  ## RNAseq data analysis
  Tutorial series for RNAseq data analysis
feature_image: "assets/genvis-dna-bg_optimized_v1a.png"
image: "assets/genvis-dna-bg_optimized_v1a.png"
layout: "home"
---


This Resource is under active development, all course material is available on github while development proceeds: [https://github.com/griffithlab/rnaseq_tutorial/wiki](https://github.com/griffithlab/rnaseq_tutorial/wiki).

Welcome to the RNA-seq Tutorial. Use the [course](/course) page to navigate your way through all exercises. Each page has a [course](/course) link at the top to bring you back to the table of contents. 

The age of affordable massively parallel sequencing has exponentially increased the availability of transcriptome profiling. RNA sequencing (RNA-seq) has rapidly become the assay of choice for interrogating RNA transcript abundance and diversity. We have developed this course to provide an introduction to RNA-seq data analysis concepts followed by integrated tutorials demonstrating the use of popular RNA-seq analysis packages. The tutorials are designed as self-contained units that include example data (Illumina paired-end RNA-seq data) and detailed instructions for installation of all required bioinformatics tools (HISAT, StringTie, etc.). Participants will gain practical experience and skills to be able to:

* Perform command-line Linux based analysis on the cloud
* Assess quality of RNA-seq data
* Align RNA-seq data to a reference genome
* Estimate known gene and transcript expression
* Perform differential expression analysis
* Discover novel isoforms
* Visualize and summarize the output of RNA-seq analyses in R
* Assemble transcripts from RNA-Seq data.

A version of this workshop was accompanied by a publication. If you find the materials here  useful, please cite the following: 
Malachi Griffith*, Jason R. Walker, Nicholas C. Spies, Benjamin J. Ainscough, Obi L. Griffith*. 2015. [Informatics for RNA-seq: A web resource for analysis on the cloud](http://dx.doi.org/10.1371/journal.pcbi.1004393). PLoS Comp Biol. 11(8):e1004393. *To whom correspondence should be addressed: E-mail: mgriffit[AT]genome.wustl.edu, ogriffit[AT]genome.wustl.edu

These materials have been developed with the support of AWS Cloud Credits for Research, the Canadian Bioinformatics Workshops, and Cold Spring Harbor Laboratory.

This workshop have been delivered many times and continues to evolve. A complete list of these past workshops is available [here](Lectures).

<table>
<tr>
<td><a href="https://aws.amazon.com/"><img src="assets/aws-powered-by.png" alt="Powerd by AWS Cloud Computing"></a></td>
<td><a href="https://bioinformatics.ca/"><img src="assets/bioinformatics_LOGO.png" alt="Canadian Bioinformatics Workshops"></a></td>
<td><a href="http://www.cshl.edu/"><img src="assets/Cold_Spring_Harbor_Laboratory_logo.png" alt="Cold Spring Harbor Laboratory"></a></td>
</tr>
</table>

