---
feature_text: |
  ## RNA-seq Bioinformatics
  Introduction to bioinformatics for RNA sequence analysis
title: Prerequisites
categories:
    - Module-00-Setup
feature_image: "assets/genvis-dna-bg_optimized_v1a.png"
date: 0000-05-01
---

## Do these before coming to the workshop:

### 1.) Command line tutorial
You are expected to be familiar with basic command line usage before the workshop. If you are unfamiliar with **the command line**, below are some suggested tutorials. If you have a Mac computer you can use the "Terminal" application to complete these tutorials. If you have access to a linux server at your institution you could also do them there. If neither of those is available you can try a free online terminal application that runs in your browser such as [sandbox.io](https://sandbox.bio/tutorials/playground) (also includes several relevant tutorials). Work through at least some of the below options.

* [Unix Intro Tutorial](https://rnabio.org/module-00-setup/0000/08/01/Unix/)
* [Unix cheat sheet](https://www.guru99.com/linux-commands-cheat-sheet.html)
* [More tutorials from BioStars](https://www.biostars.org/p/9572411/)

### 2.) Introduction to R tutorial
Knowing R is critical to this workshop. Using R installed on your computer, please work through at least some of the following tutorials.

* [R Tutorial (W3 Schools)](https://www.w3schools.com/r/default.asp)
* [Introduction to R (datacamp)](https://app.datacamp.com/learn/courses/free-introduction-to-r) (Take first free chapter)
* [R Review & Assessment](https://nceas.github.io/oss-lessons/r-review-and-assessment/r-review-and-assessment.html)
* [GenViz Intro R tutorial](https://genviz.org/module-02-r/0002/02/01/introductionToR/) (Note this tutorial is more advanced, especially the final exerxise, and not required)

### 3.) Reading materials
Please browse these articles, select a few that fill in knowledge gaps, and review before coming to the workshop:

**Some recent reviews (2024-2025)**
* Experimental design. [Technical considerations for planning an RNA-Sequencing experiment](https://doi.org/10.1186/s12864-025-12094-8).
* Spatial transcriptomics. [Review of spatial transcriptomics data alignment and integration](https://doi.org/10.1093/nar/gkaf536).
* Bioinformatics for transcriptomics. [Bioinformatics perspectives on transcriptomics: A comprehensive review of bulk and single-cell RNA sequencing analyses](https://doi.org/10.1002/qub2.78).
* Single-cell + long reads. [Advances in single-cell long-read sequencing technologies](https://doi.org/10.1093/nargab/lqae047).
* Deep learning. [Deep learning in single-cell and spatial transcriptomics data analysis: advances and challenges from a data science perspective](https://doi.org/10.1093/bib/bbaf136).

**Some slightly older but excellent reviews (2019-2023)**
* RNA-seq data fundamentals. [RNA Sequencing: The Teenage Years](https://doi.org/10.1038/s41576-019-0150-2).
* RNA-seq analysis and statistics fundamentals. [RNA Sequencing Data: Hitchhikerâ€™s Guide to Expression Analysis](https://doi.org/10.1093/bib/bbac529).
* RNA-seq analysis for tumor immunity. [Tutorial: integrative computational analysis of bulk RNA-sequencing data to characterize tumor immunity using RIMA](https://doi.org/10.1038/s41596-023-00841-8).
* scRNA-seq technologies and applications. [Single-cell RNA sequencing technologies and applications: A brief overview](https://doi.org/10.1002/ctm2.694).
* DNA sequencing and variant calling overview [Novel sequencing technologies and bioinformatic tools for deciphering the non-coding genome](https://doi.org/10.1515/medgen-2021-2072).

Optional readings:

The above are just a few published examples that we find particularly good. We have collected a much more extensive list of resources below. Simply take a look at these and note for future reference. Note, these are focused primary towards RNAseq analysis but are often more generally applicable. 

* [RNA-seq resources](https://rnabio.org/resources/)
