---
feature_text: |
  ## RNA-seq Bioinformatics
  Introduction to bioinformatics for RNA sequence analysis
title: Introduction to Single-cell RNA-seq
categories:
    - Module-08-scRNA
feature_image: "assets/genvis-dna-bg_optimized_v1a.png"
date: 0008-01-01
---

### Module 7 - Key concepts
* Key differences between bulk RNA-seq data and single-cell RNA-seq data
* Technology, sensitivity, and throughput of plate-based and droplet-based scRNA-seq platforms
* 10xGenomics Chromium technology, library structure, and Cellranger analysis pipeline
* Key metrics of data quality
* Genomic composition of scRNA-seq data
* Statistical properties of key variables (genes per cell, UMIs per cell, etc.)
* Downstream analysis of the feature-barcode (i.e. gene-by-cell) matrix
* Rationale underlying filtering, normalization, and other quality control steps
* Intuition and interpretation of dimensionality reduction techniques
* Interpretation of 2-dimensional data layouts generated by the t-SNE and UMAP algorithms
* Rationale for cell clustering and cluster interpretation
* Further strategies for cell type inference and data interpretation

### Module 7 - Learning objectives
* Understand the applications of scRNA-seq
* Learn the advantages of different scRNA-seq platforms
* Understand the 10xGenomics technology
* Learn the Cellranger commands for initial processing of 10xGenomics data
* Understand Cellranger output files
* Learn to assess the quality of your data 
* Use the loupe browser to perform initial data exploration
* Understand the rationale underlying scRNAseq analysis steps
* Understand biological interpretation of scRNA-seq data
* Learn how to perform custom analyses of scRNAseq data using R
* Implement a complete scRNAseq analysis pipeline in R

### Module 8 - Lecture

* [scRNAseq_Part1.pdf](https://github.com/griffithlab/rnabio.org/raw/master/assets/module_8/CBW2020_scRNAseq_1_APetti.pdf)
* [scRNAseq_Part2.pdf](https://github.com/griffithlab/rnabio.org/raw/master/assets/module_8/CBW2020_scRNAseq_2_APetti.pdf)

